apicast/src/oauth/keycloak.lua

```diff
@@ -0,0 +1,276 @@
+    http_client = http_client,
+    service = service
+    }, mt)
+end
+
+function _M.respond_and_exit(status, body, headers)
+  -- TODO: is there a better way to populate the response headers?..
+  if headers then
+    for name,value in pairs(headers) do
+      ngx.header[name] = value
+    end
+  end
+
+  ngx.status = status
+  ngx.print(body)
+  ngx.exit(ngx.HTTP_OK)
+end
+
+function _M.respond_with_error(status, message)
+
+  --TODO: as per the RFC (https://tools.ietf.org/html/rfc6749#section-5.2) return WWW-Authenticate response header if 401
+  local headers = {
+    ['Content-Type'] = 'application/json;charset=UTF-8'
+  }
+  local err_msg = { error = message }
+  local body = cjson.encode(err_msg)
+  _M.respond_and_exit(status, body, headers)
+end
+
+function _M.authorize_check_params(params)
+  local response_type = params.response_type
+  local required_params = _M.params.response_type
+  if not response_type then return false, 'invalid_request' end
+  if not required_params[response_type] then return false, 'unsupported_response_type' end
+
+  for _,v in ipairs(required_params[response_type]) do
+    if not params[v] then
+      return false, 'invalid_request'
+    end
+  end
+
+  return true
+end
+
+function _M.token_check_params(params)
+  local grant_type = params.grant_type
+  local required_params = _M.params.grant_type
+  if not grant_type then return false, 'invalid_request' end
+  if not required_params[grant_type] then return false, 'unsupported_grant_type' end
+
+  for _,v in ipairs(required_params[grant_type]) do
+    if not params[v] then
+      return false, 'invalid_request'
+    end
+  end
+  return true
+end
+
+-- Parses the token - in this case we assume it's a JWT token
+-- Here we can extract authenticated user's claims or other information returned in the access_token
+-- or id_token by RH SSO
+local function parse_and_verify_token(self, jwt_token)
+  local jwt_obj = jwt:verify(self.config.public_key, jwt_token)
+
+  if not jwt_obj.verified then
+    local err = "[jwt] failed verification for token: "..jwt_token.." reason: "..jwt_obj.reason
+    ngx.log(ngx.INFO, err)
+    return jwt_obj, err
+  end
+  return jwt_obj
+end
+
+function _M.credentials(self, access_token)
+  local jwt_obj, err = parse_and_verify_token(self, access_token)
+
+  if not jwt_obj then
+    err = "[jwt] failed to parse token: "..access_token
+    return nil, err
+  else
+    if jwt_obj.payload then
+      local app_id = jwt_obj.payload.aud
+
+      ------
+      -- oauth credentials for keycloak
+      -- @field 1 Client id
+      -- @field app_id Client id
+      -- @table credentials_oauth
+      return { app_id, app_id = app_id }, err
+    else
+      err = "[jwt] failed to parse token: "..jwt_obj.reason
+      return nil, err
+    end
+  end
+end
+
+function _M.check_credentials(self, params)
+  local backend = backend_client:new(self.service)
+
+  local args = {
+      app_id = params.client_id,
+      app_key = params.client_secret,
```

Comment:
Just to clarify, this is only used to check credentials for the purpose of deciding whether or not to send the request on to keycloak during the OAuth handshake, not for API request validation